WalletWasabi/WabiSabi/Client/CredentialDependencies/CredentialEdgeSet.cs

```diff
@@ -0,0 +1,253 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Client.CredentialDependencies
+{
+	public record CredentialEdgeSet
+	{
+		public CredentialType CredentialType { get; init; }
+		public ImmutableDictionary<RequestNode, ImmutableHashSet<CredentialDependency>> Predecessors { get; init; } = ImmutableDictionary.Create<RequestNode, ImmutableHashSet<CredentialDependency>>();
+		public ImmutableDictionary<RequestNode, ImmutableHashSet<CredentialDependency>> Successors { get; init; } = ImmutableDictionary.Create<RequestNode, ImmutableHashSet<CredentialDependency>>();
+		public ImmutableDictionary<RequestNode, long> EdgeBalances { get; init; } = ImmutableDictionary.Create<RequestNode, long>();
+
+		public long Balance(RequestNode node) => node.InitialBalance(CredentialType) + EdgeBalances[node];
+
+		public ImmutableHashSet<CredentialDependency> InEdges(RequestNode node) => Predecessors[node];
+
+		public ImmutableHashSet<CredentialDependency> OutEdges(RequestNode node) => Successors[node];
+
+		public int InDegree(RequestNode node) => InEdges(node).Count;
+
+		public int OutDegree(RequestNode node) => OutEdges(node).Where(x => x.Value != 0).Count();
+
+		public int ZeroOnlyOutDegree(RequestNode node) => OutEdges(node).Where(x => x.Value == 0).Count();
+
+		public int RemainingInDegree(RequestNode node) => node.MaxInDegree - InDegree(node);
+
+		public int RemainingOutDegree(RequestNode node) => node.MaxOutDegree - OutDegree(node);
```

Comment:
This methods (among others) return integer but they are never used to perform any math, instead they are used to decide what to do `if (RemainingInDegree(whatever) == 0) then...`. I think in that case they could return a boolean and start ttheir names with `Is`/`Are` as `bool AreRemainingInDegree(RequestNode)`

Reply:
it's also used for ordering, but I think we might be able to implement this suggestion anyway

Reply:
So, there's three conditions we actually care about:

1. is the remaining in/out degree == 0
2. is the remaining in/out degree == 1
3. is the remaining in/out degree > 1 (or in our case, degree == 2)

the second condition is needed because when that is the case, although an edge can be added from that node, the value of the edge is constrained to be exactly the remaining balance.

I think we can model this with two methods that return bool, but I'm not sure what to call them.

A similar interaction then happens with the zero credentials, for K=2 if the node has 1 edge with a value equal to its entire positive balance, then this node now has 3 remaining zero credentials.